doc/how_to/migrate/from_streamlit.md

```diff
@@ -0,0 +1,1011 @@
+    ax = fig.subplots()
+    ax.hist(data, bins=bins)
+    return fig
+
+pn.extension(sizing_mode="stretch_width", template="bootstrap")
+
+data = get_data()
+bins = pn.widgets.IntSlider(value=20, start=10, end=30, step=1)
+bplot = pn.bind(plot, data, bins)
+pn.Column(bins, pn.panel(bplot, loading_indicator=True)).servable()
+```
+
+![Panel Cache Example](https://assets.holoviz.org/panel/gifs/panel_cache_example.gif)
+
+You can also use `pn.cache` as an function. I.e. as
+
+```python
+plot = pn.cache(plot)
+```
+
+Using `pn.cache` as a function can help you keep your business logic
+(`data` and `plot` function) and your caching logic (when and how to apply caching) separate. This
+can help you reusable and maintainable code.
+
+### Typical Caching Mistakes
+
+- Using `pn.state.cache` instead of `pn.cache`. `pn.state.cache` is a simple dictionary and cannot be used as a function or a decorator.
+
+### Cache Migration Steps
+
+To migrate
+
+- replace `st.cache_data` and `st.cache_resource` with `pn.cache`.
+
+You should consider caching expensive
+
+- functions that are run when your page loads
+- *bound functions*
+
+## Multi Page Apps
+
+Migrating your Streamlit multi page app to Panel is simple. In Panel each page is simply a file
+that you *serve*
+
+```bash
+panel serve home.py page1.py page2.ipynb
+```
+
+You can specify the *home* page with the `--index` flag.
+
+```bash
+panel serve home.py page1.py page2.ipynb --index=home
+```
+
+## Session State
+
+Session state (and callbacks) were added to Streamlit as an after thought
+to enable sharing variables between re-runs and across apps inside a multipage app.
+
+Panels execution model is very different and session state is not really a topic.
+
+### Streamlit Session State Example
+
+```python
+import random
+from uuid import uuid4
+
+import streamlit as st
+
+st.session_state.setdefault("rows", {})
+
+def create_row(hex_id=None):
+    if hex_id is None:
+        hex_id = uuid4().hex
+        value = random.randint(0, 100)
+        st.session_state.rows[hex_id] = value
+    else:
+        value = st.session_state.rows[hex_id]
+
+    st.number_input("Enter a number", value=value, key=f"num_{hex_id}")
+
+st.title("Dynamically add new rows")
+
+button_cols = st.columns(2)
+with button_cols[0]:
+    add_row = st.button("Add new row", use_container_width=True)
+with button_cols[1]:
+    clear_rows = st.button("Clear all rows", use_container_width=True)
+
+if clear_rows:
+    st.session_state.rows = {}
+
+for hex_id in st.session_state.rows.keys():
+    create_row(hex_id=hex_id)
+
+if add_row:
+    create_row()
+```
+
+![Streamlit Session State Example](https://user-images.githubusercontent.com/42288570/248444403-3e9548a5-71a6-4154-8e85-b9d9e08b9111.png)
+
+With Streamlit you need session state to keep track of the dynamically created `rows`.
+
+### Panel Session State Example
+
+```python
+import random
+
+import panel as pn
+pn.extension(sizing_mode="stretch_width")
+
+def create_row(_):
```

Comment:
Would make Panel buttons so much easier to explain and use if they could take callbacks without arguments too @philippjfr